Add metadata for torch jit TracedModules.

zdevito opened the pull request:
Summary:
I've extended our model metadata framework in this diff to support
traced modules as well. Re-used a lot of components from the previous
implementation of ScriptModule metadata.

Tracing is a little different from Scripting since you can't just create a
subclass of TopLevelTraceModule (type returned by torch.jit.trace) and attach
metadata the way we did for ScriptModule. As a result, I've introduced a
separate API torch.fb.jit_trace which returns an instance of
TracedModuleWithMetadata which is a subclass of TopLevelTracedModule. As a
result, we can now attach metadata to this instance.

Differential Revision: D14117966

houseroad commented on torch/jit/__init__.py:
```diff
@@ -577,7 +577,6 @@ def ignore_lib_warnings():
 TracerWarning.ignore_lib_warnings()
 torch._C._tracer_warn_use_python()
 
-
```
lint?

houseroad commented on torch/jit/__init__.py:
```diff
@@ -1404,6 +1382,38 @@ def forward(self, *args, **kwargs):
         return self._get_method('forward')(*args, **kwargs)
 
 
+def _trace(func,
```
shall we just put the `_trace` after `trace`?

pritamdamania committed:
Add metadata for torch jit TracedModules. (#17311)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/17311

I've extended our model metadata framework in this diff to support
traced modules as well. Re-used a lot of components from the previous
implementation of ScriptModule metadata.

Tracing is a little different from Scripting since you can't just create a
subclass of TopLevelTraceModule (type returned by torch.jit.trace) and attach
metadata the way we did for ScriptModule. As a result, I've introduced a
separate API torch.fb.jit_trace which returns an instance of
TracedModuleWithMetadata which is a subclass of TopLevelTracedModule. As a
result, we can now attach metadata to this instance.

Differential Revision: D14117966

fbshipit-source-id: 2b63fd8053046dde3e0d21118a2e7337f45572a5
